mllib/src/test/java/org/apache/spark/ml/feature/JavaPCASuite.java

```diff
@@ -67,7 +66,7 @@ public void testPCA() {
     JavaRDD<Vector> dataRDD = jsc.parallelize(points, 2);
 
     RowMatrix mat = new RowMatrix(dataRDD.map(
-        (Vector vector) -> (org.apache.spark.mllib.linalg.Vector) new DenseVector(vector.toArray())
+        (Vector vector) -> org.apache.spark.mllib.linalg.Vectors.fromML(vector)
```

Comment:
This is an interesting point; what if the data is a mix of sparse and dense? At least this test now covers that case a bit.

The check above only tests the first element, but that's the only reasonable thing to do. If it were, for example, mostly sparse but the first happened to be dense, then this change could be problematic. I think it's OK to assume that most real-world uses are all sparse or dense. It is possible that some data in a sparse case is occasionally encoded as sparse as an optimization if it has a lot of zeroes, but this case works fine here.

Reply:
Sure, as you said, if the **first item** is sparse vector, it will align with computeSparseVectorCovariance logic, and if the **first item** is a dense vector, it will align with computeDenseVectorCovariance logic.  The rest is user's choice. 
But, maybe we can add some notes into the annotation of function computeCovariance, 
give user some notes, like:

/**
   * Computes the covariance matrix, treating each row as an observation.
   *
   * Note:
   * When the first row is DenseVector, we use the computeDenseVectorCovariance
   * to calculate the covariance matrix, and if the first row is SparseVector, we
   * use the computeSparseVectorCovariance to calculate the covariance matrix
   *
   * @return a local dense matrix of size n x n
   *
   * @note This cannot be computed on matrices with more than 65535 columns.
   */
  @Since("1.0.0")
  def computeCovariance(): Matrix = {

Reply:
I don't think that helps a caller much as they don't know what the implication of this is. I'd leave it for now.